### PR TITLE
create/get keys with hashed keys logic added

### PR DESCRIPTION
**Pull request by dencoded**
added changes for https://github.com/TykTechnologies/tyk/issues/1256

- creating key - `POST /keys/create`, `POST /keys` and `POST /keys/{keyName}` also return field `"key_hash"` if key hashing is switched on
- get all (or per API) keys  - `GET /keys` returns list of key hashes if key hashing is switched on
- get one key info by hash `GET /keys/{keyName}?hashed=true"`
- `DELETE /keys/{keyName}?hashed=true"` - remains the same (delete by hash)

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -620,41 +638,22 @@ func handleUpdateHashedKey(keyName, apiID, policyId string) (interface{}, int) {
...
-
-		return apiError("Unmarshalling failed"), 400
-	}
-
 	// Set the policy
 	sess.LastUpdated = strconv.Itoa(int(time.Now().Unix()))
 	sess.SetPolicies(policyId)
 
-	sessAsJS, err := json.Marshal(sess)
+	err := sessionManager.UpdateSession(keyName, &sess, 0)
```
Comment (buger) [inline api.go]: I noticed you not passing any info that key is hashed here, are you sure it will not do double hashing here?

File: api.go [hunk 2]
```diff
@@ -743,7 +735,7 @@ func handleOrgAddOrUpdate(keyName string, r *http.Request) (interface{}, int) {
 		rawKey := QuotaKeyPrefix + storage.HashKey(keyName)
 
 		// manage quotas separately
-		DefaultQuotaStore.RemoveSession(rawKey)
+		DefaultQuotaStore.Store().DeleteRawKey(rawKey)
```
Comment (buger) [inline api.go]: Why this change?

**Comment by buger** [comment]
Looks good, pls add new config variable (false by default), which enable **listing** of hashed keys. There are case, like our cloud when you have lot of keys, when you want to disable listing, but still allow querying hashed keys.

File: api.go (hunk 1 above)
Comment (dencoded) [reply api.go]: thats true, looks like I missed it. I was experimenting with `UpdateSession` and then decided to leave it as is because it is used a lot. I will push a fix for it

**Comment by dencoded** [comment]
one more thing for documenting:
- added new config value `"enable_hashed_keys_listing"` with `"false"` by default

this config value controls access to `GET /keys` when keys are hashed.
- so, if `"hash_keys":true` and `"enable_hashed_keys_listing":false` then `GET /keys` will return 404 with explanation why
- if `"hash_keys":true` and `"enable_hashed_keys_listing":true` then `GET /keys` will return 200 with list of hashes
